frontend/src/metabase/components/AppBanner/AppBanner.tsx

```diff
@@ -28,6 +40,28 @@ export const AppBanner = () => {
     return <ReadOnlyBanner />;
   }
 
+  if (isValidTrial) {
+    const daysRemaining = dayjs(tokenExpiryTimestamp).diff(
+      getCurrentUTCTimestamp(),
+      "days",
+    );
+
+    const showBanner = shouldShowTrialBanner({
+      tokenExpiryTimestamp,
+      daysRemaining,
+      lastDismissed,
+    });
+
+    return (
+      showBanner && (
+        <TrialBanner
+          daysRemaining={daysRemaining}
+          onClose={() => setLastDismissed(getCurrentUTCTimestamp())}
+        />
+      )
+    );
```

Comment:
I think you want to put this return statement in another `if` statement. This logic short-circuits the possibility of showing the payment banner to any trial users.. and is harder to understand than just falling through to the `null` return at the end of the function

```suggestion
    if(showBanner) {
        return (
          <TrialBanner
            daysRemaining={daysRemaining}
            onClose={() => setLastDismissed(getCurrentUTCTimestamp())}
          />
      );
    }

```

Reply:
Payment banner and Trial banner are mutually exclusive. One is shown to the self-hosted instances and the other one is shown to the hosted/cloud instances.

But I can still reorder this if it makes it easier to read.